mitodl/ocw-to-hugo
# Handle archived courses

#### Pre-Flight checklist


#### What are the relevant tickets?
Part of #274 

#### What's this PR do?
Adds markdown about archived courses to the course json data. Archived courses are found by iterating through the course uids in `is_update_of` and parsing the link from `dspace_handle` if it exists. This is done recursively because courses [like this one](https://ocw.mit.edu/courses/political-science/17-40-american-foreign-policy-past-present-and-future-fall-2017/) for example list archived courses which are anywhere in the parent tree. 

This also refactors the pathLookup code a bit so that `byMasterSubject` (renamed to `coursesByMasterSubject`) uses course info in `byUid` instead of storing it in `byMasterSubject`. 

There are 5 commits which are broken up so the reviewer can differentiate between the refactoring steps and the new archived courses feature. All tests should be passing for each commit so it should show that the refactoring didn't change output at any point (hopefully).

#### How should this be manually tested?
First, convert all raw courses to parsed courses using the latest version of ocw-data-parser (`release` is fine). There are some fields which are necessary for ocw-to-hugo to have information needed to construct archive links.

Convert all courses and compare the directories for the output. You should see a new field `archived_courses` in `1-011-project-evaluation-spring-2011` with a list like this:

    [1.011 PROJECT EVALUATION](https://dspace.mit.edu/handle/1721.1/75001) |  SPRING 2005
    [1.011 PROJECT EVALUATION](https://dspace.mit.edu/handle/1721.1/36392) |  SPRING 2003

If you view the [course on the legacy site](https://ocw.mit.edu/courses/civil-and-environmental-engineering/1-011-project-evaluation-spring-2011/) you should see the same two links to the same two courses. They might be in a different order.

Note that this PR increases the running time from 3m:0s to 3m:20s minutes since the refactoring means it takes multiple passes to build the paths. I think that's fine personally since we are trying to minimize memory usage, not running time necessarily, but let me know if you have any thoughts there
